front/src/common/IntervalsEditor/IntervalsEditor.tsx

```diff
@@ -1,191 +1,151 @@
-import { BsFillTrashFill } from 'react-icons/bs';
-import { IoIosAdd, IoIosSave } from 'react-icons/io';
-import React, { useState, useRef, useCallback, useMemo } from 'react';
-import { LinearMetadataDataviz } from 'common/IntervalsDataViz/dataviz';
-
-import { TbZoomIn, TbZoomOut, TbZoomCancel, TbScissors, TbArrowsHorizontal } from 'react-icons/tb';
+import React, { useState, useRef, useMemo, useEffect } from 'react';
+import { isNil, max } from 'lodash';
 
 import {
   LinearMetadataItem,
-  createSegmentAt,
-  fixLinearMetadataItems,
-  getLineStringDistance,
+  createEmptySegmentAt,
   getZoomedViewBox,
   removeSegment,
   resizeSegment,
   splitAt,
   transalteViewBox,
 } from 'common/IntervalsDataViz/data';
+import { LinearMetadataDataviz } from 'common/IntervalsDataViz/dataviz';
+import { tooltipPosition } from 'common/IntervalsDataViz/utils';
 
-import { FieldProps } from '@rjsf/core';
-import { isNil, max as fnMax, cloneDeep } from 'lodash';
-
-import { tooltipPosition, notEmpty } from 'common/IntervalsDataViz/utils';
-import { ValueOf } from 'utils/types';
-import InputSNCF from 'common/BootstrapSNCF/InputSNCF';
-import { useTranslation } from 'react-i18next';
-import SelectSNCF from 'common/BootstrapSNCF/SelectSNCF';
 import IntervalsEditorTootlip from './IntervalsEditorTooltip';
+import IntervalsEditorCommonForm from './IntervalsEditorCommonForm';
+import {
+  INTERVALS_EDITOR_TOOLS,
+  INTERVAL_TYPES,
+  IntervalItem,
+  IntervalsEditorTool,
+  IntervalsEditorToolsConfig,
+} from './types';
+import ZoomButtons from './ZoomButtons';
+import ToolButtons from './IntervalsEditorToolButtons';
+import IntervalsEditorMarginForm from './IntervalsEditorMarginForm';
+import IntervalsEditorSelectForm from './IntervalsEditorSelectForm';
 
-export const TOOLS = Object.freeze({
-  cutTool: Symbol('cutTool'),
-  deleteTool: Symbol('deleteTool'),
-  translateTool: Symbol('translateTool'),
-  addTool: Symbol('addTool'),
-});
-
-type IntervalsEditorParams = {
-  cutTool?: boolean;
-  deleteTool?: boolean;
-  translateTool?: boolean;
-  addTool?: boolean;
+type IntervalsEditorProps = {
+  defaultValue: number | string;
+  setData: (newData: IntervalItem[]) => void;
   showValues?: boolean;
-};
-
-type IntervalsEditorMetaProps = {
-  params?: IntervalsEditorParams;
-  valueField: string;
-  values?: string[];
-  defaultValue?: number;
-  units?: string[];
-  defaultUnit?: string;
-  onChange: (newData: LinearMetadataItem[]) => void;
-};
+  title?: string;
+  toolsConfig?: IntervalsEditorToolsConfig;
+  totalLength: number;
+} & (
+  | {
+      intervalType: INTERVAL_TYPES.NUMBER_WITH_UNIT;
+      data: LinearMetadataItem<{ value: number | string; unit: string }>[];
+      defaultUnit?: string;
+      fieldLabel: string;
+      units: string[];
+    }
+  | {
+      intervalType: INTERVAL_TYPES.SELECT;
+      data: LinearMetadataItem<{ value: number | string }>[];
+      selectOptions: string[];
+    }
+);
```

Comment:
Not sure how to implement it but to narrow the "data" type maybe there is something to dig with conditional types ?

Reply:
Yes, I tried to do it, but it is quite tricky :( I'm going to ask Benoît or Alexis for this, but I think it should wait a future PR